src/main/java/com/github/shyiko/mysql/binlog/BinaryLogClient.java

```diff
@@ -962,19 +964,37 @@ private void updateClientBinlogFilenameAndPosition(Event event) {
         }
     }
 
+    private EventData unwrapEventData(EventData eventData) {
+        if (eventData instanceof EventDeserializer.EventDataWrapper) {
+            return ((EventDeserializer.EventDataWrapper) eventData).getInternal();
+        } else {
+            return eventData;
+        }
+    }
+
+    private void addGtidToSet(String gtid) {
+        if ( gtid == null )
+            return;
+
+        synchronized (gtidSetAccessLock) {
+            gtidSet.add(currentGtid);
+        }
+    }
+
     private void updateGtidSet(Event event) {
         EventHeader eventHeader = event.getHeader();
         if (eventHeader.getEventType() == EventType.GTID) {
-            synchronized (gtidSetAccessLock) {
-                if (gtidSet != null) {
-                    EventData eventData = event.getData();
-                    GtidEventData gtidEventData;
-                    if (eventData instanceof EventDeserializer.EventDataWrapper) {
-                        gtidEventData = (GtidEventData) ((EventDeserializer.EventDataWrapper) eventData).getInternal();
-                    } else {
-                        gtidEventData = (GtidEventData) eventData;
-                    }
-                    gtidSet.add(gtidEventData.getGtid());
+            GtidEventData gtidEventData = (GtidEventData) unwrapEventData(event.getData());
+            currentGtid = gtidEventData.getGtid();
+        } else if ( gtidSet != null ) {
```

Comment:
gtidSet is no longer protected by gtidSetAccessLock

Reply:
it is, in `addGtidToSet`